### PR TITLE
Support for tiler to render fewer plots on large maps

### PR DESCRIPTION
**Pull request by RickMohr**
When viewing a zoomed-out tree map, multiple plots in a local area are rendered on top of one another. We set `MapFeature.hide_at_zoom` so the tiler can render just one plot in such cases, speeding up tile creation for maps with many plots.

The plan is to recompute `hide_at_zoom` nightly. After a large tree import, tiling could be somewhat slower until the next day.

To test:

1. Run migrations
2. Run management command to set `hide_at_zoom` for all plots in a large instance, e.g. `manage.py recompute_hide_at_zoom latreemap`. (It takes several minutes to complete.)
3. Load the tree map (with dev tools open to disable cacheing) and click around
4. To compare performance, do an advanced search of "Last Updated" from 5 years ago to now. (When a filter is passed to the tiler the optimizations aren't used.)

Requires OpenTreeMap/otm-tiler#93
Connects #2374

### CODE REVIEW

**Comment by RickMohr** [comment]
Without `hide_at_zoom`:
image

With `hide_at_zoom`:
image

**Comment by jwalgran** [comment]
When I tried to move a plot through the UI I got a permission exception

```
AuthorizeException at /latreemap/features/1099132/
Can't edit field hide_at_zoom on Plot
```

We need `hide_at_zoom` to be exempt from permission checks.

Full trace:

```
Traceback:
File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/opt/app/core/opentreemap/treemap/decorators.py" in wrapper
  45.             return view_fn(request, instance, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in routed
  35.             return kwargs['ELSE'](request, *args2, **kwargs2)
File "/opt/app/core/opentreemap/treemap/decorators.py" in wrapper
  169.             return view_fn(request, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in newreq
  125.                 outp = req_function(request, *args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in newreq
  106.         outp = req_function(request, *args, **kwargs)
File "/opt/app/core/opentreemap/treemap/decorators.py" in wrapper
  193.         return view_fn(request, instance, *args, **kwargs)
File "/opt/app/core/opentreemap/treemap/decorators.py" in run_and_catch_validations
  144.             return req(*args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py" in routed
  33.             return req_method(request, *args2, **kwargs2)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in update_map_feature_detail
  152.     return _request_to_update_map_feature(request, feature)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in _request_to_update_map_feature
  42.     feature, tree = update_map_feature(request_dict, request.user, feature)
File "/usr/local/lib/python2.7/dist-packages/django/utils/decorators.py" in inner
  145.                     return func(*args, **kwargs)
File "/opt/app/core/opentreemap/treemap/views/map_feature.py" in update_map_feature
  294.         update_hide_at_zoom_after_move(feature, user, old_geom)
File "/opt/app/core/opentreemap/treemap/lib/hide_at_zoom.py" in update_hide_at_zoom_after_move
  116.         feature.save_with_user(user)
File "/opt/app/core/opentreemap/treemap/models.py" in save_with_user
  603.         super(MapFeature, self).save_with_user(user, *args, **kwargs)
File "/opt/app/core/opentreemap/treemap/udf.py" in save_with_user
  1217.         super(UDFModel, self).save_with_user(user, *args, **kwargs)
File "/opt/app/core/opentreemap/treemap/audit.py" in save_with_user
  861.                                             (field, self._model_name))

Exception Type: AuthorizeException at /latreemap/features/1099132/
Exception Value: Can't edit field hide_at_zoom on Plot
```

**Comment by maurizi** [comment]
> We need hide_at_zoom to be exempt from permission checks.

We did this for `MapFeature.updated_at` in a way that should be very easy to reuse: https://github.com/OpenTreeMap/otm-core/blob/master/opentreemap/treemap/models.py#L547

**Comment by RickMohr** [comment]
@jwalgran, fixed. @maurizi, thanks very much for the pointer.

File: opentreemap/treemap/lib/hide_at_zoom.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+MAX_ZOOM = 14
+MIN_ZOOM = 0
+
+# When viewing a zoomed-out tree map, multiple plots in a local area are
+# rendered on top of one another. We set MapFeature.hide_at_zoom so the tiler
+# can render just one plot in such cases, speeding up tile creation for maps
+# with many plots.
+#
+# At each zoom level Z we find groups of plots located in the same cell of
+# a 2-pixel grid, and set hide_at_zoom = Z for all but one plot in each group.
```
Comment (jwalgran) [inline opentreemap/treemap/lib/hide_at_zoom.py]: We may want this comment to refer to a  "GRID_PIXELS sized grid"

Is there a reason why we have chosen `GRID_PIXELS = 2`?

Comment (RickMohr) [inline opentreemap/treemap/lib/hide_at_zoom.py]: Good point. I tried a bunch of sizes. The bigger the grid the better the speedup, but with larger grids the map starts to look funny. 2 pixels was the sweet spot.

**Comment by jwalgran** [comment]
I did see an 8-10x speedup of tile response times on this branch.

I tested moving and deleting plots and ran DB queries to verify that `hide_at_zoom` values were being updated.

:+1:

**Pull request merged**